app/code/Vladimirl/Chatter/Controller/Submit/Submit.php

```diff
@@ -57,17 +98,56 @@ public function generateHash(): string
     public function execute()
     {
         try {
-            $request = $this->getRequest();
-            $websiteId = (int)$this->storeManager->getWebsite()->getId();
-            $authorType = 'customer';
-            $authorName = 'anonymous';
+            if (!$this->formKeyValidator->validate($this->getRequest())) {
+                throw new LocalizedException(__('Something went wrong!'));
+            }
+
+            $customerId = (int) $this->customerSession->getId();
+            $websiteId = (int) $this->storeManager->getWebsite()->getId();
+
+            if (!$chatHash = $this->customerSession->getChatHash()) {
+                $this->customerSession->setChatHash($this->generateHash());
+            }
+
+            if ($customerId) {
+                $transaction = $this->transactionFactory->create();
+                $authorType = 'customer';
+                $authorName = $this->customerSession->getCustomerData()->getEmail();
+
+                $messageCollection = $this->messageCollectionFactory->create();
+                $messageCollection->addCustomerIdFilter($customerId)
+                    ->addAuthorTypeFilter($authorType);
+                $oldChatHash = $messageCollection->getFirstItem()->getChatHash();
+                if ($oldChatHash === null) {
+                    $oldChatHash = $this->customerSession->getChatHash();
+                }
+
+                $messageCollection = $this->messageCollectionFactory->create();
+                $messageCollection->addChatHashFilter($chatHash);
+
+                foreach ($messageCollection as $existingMessage) {
+                    if ($existingMessage->getAuthorId() !== $customerId) {
+                        $existingMessage->setAuthorType($authorType)
+                            ->setAuthorId($customerId)
+                            ->setAuthorName($authorName)
+                            ->setChatHash($oldChatHash);
+                    }
+                    $transaction->addObject($existingMessage);
+                }
+                $transaction->save();
+                $this->customerSession->setChatHash($oldChatHash);
+
+            } else {
+                $authorType = 'unknown';
```

Comment:
@Vladimir-L Author type is probably "guest" or something like that. This is always guest if the message is posted from the fromend.